src/bodse/scheduler.py

```diff
@@ -0,0 +1,704 @@
+# -*- coding: utf-8 -*-
+"""
+Functionality for automating the downloading and processing of BODS data.
+
+The scheduler will perform the following tasks on a regular schedule:
+- Download the GTFS schedule from BODS
+- Download AVL data from BODS for a set period
+- Calculate an adjusted GTFS schedule using AVL data
+- Upload all datasets to a PostgreSQL database
+"""
+
+##### IMPORTS #####
+
+# Built-Ins
+import calendar
+import datetime
+import enum
+import logging
+import pathlib
+import re
+import time
+import traceback
+from collections import abc
+from typing import Any, Optional
+
+# Third Party
+import pydantic
+from caf.toolkit import config_base, log_helpers
+from pydantic import dataclasses, types
+
+# Local Imports
+import bodse
+from bodse import database, request, teams, timetable
+from bodse.avl import adjust, avl
+from bodse.avl import database as avl_db
+
+##### CONSTANTS #####
+
+LOG = logging.getLogger(__name__)
+WAIT_TIME = 3600 * 12  # 0.5 days
+
+
+##### CLASSES & FUNCTIONS #####
+
+
+class Day(enum.IntEnum):
+    """Days of the week."""
+
+    MONDAY = calendar.MONDAY
+    TUESDAY = calendar.TUESDAY
+    WEDNESDAY = calendar.WEDNESDAY
+    THURSDAY = calendar.THURSDAY
+    FRIDAY = calendar.FRIDAY
+    SATURDAY = calendar.SATURDAY
+    SUNDAY = calendar.SUNDAY
+
+    @property
+    def name_and_number(self) -> str:
+        """Return name and number of day."""
+        return f"{self.name} ({self.value})"
+
+    @classmethod
+    def _missing_(cls, value: Any) -> "Day":
+        try:
+            value = int(value)
+        except ValueError:
+            pass
+
+        if isinstance(value, int):
+            for i in cls:
+                if value == i.value:
+                    return i
+
+        else:
+            value = str(value)
+            upper = value.upper().strip()
+
+            if upper in [i.name for i in cls]:
+                return cls[upper]
+
+            for i in cls:
+                if i.name[:3] == upper:
+                    return i
+
+        raise ValueError(
+            f"{value!r} is not a valid {cls.__name__} should be one of "
+            + ", ".join(i.name_and_number for i in cls)
+        )
+
+
+@dataclasses.dataclass
+class TaskParameters:
+    """Parameters for scheduled BODSE tasks."""
+
+    gtfs_folder: pydantic.DirectoryPath
+    avl_download_time: avl.DownloadTime
+    max_timetable_age_days: int = pydantic.Field(7, ge=0)
+    max_avl_age_days: int = pydantic.Field(7, ge=0)
+    avl_database_delete_age_days: int = pydantic.Field(30, ge=0)
+
+
+class SchedulerConfig(config_base.BaseConfig):
+    """Parameters for running BODSE scheduler."""
+
+    output_folder: types.DirectoryPath
```

Comment:
A question mainly here.

is `types.DirectoryPath` interchangable with `pydantic.DirectoryPath`, noting that it would be `pydantic.tpyes.DirectoryPath` opposed to `pydantic.DirectoryPath`? 

Assming to be different from `pathlib.Path`, which can be a dir or file path, this seems to explicitly be a dir only path

Reply:
makes sense, cheers!